dubbo-test/dubbo-test-check/src/main/java/org/apache/dubbo/test/check/registrycenter/initializer/DownloadZookeeperInitializer.java

```diff
@@ -46,46 +48,105 @@
+     *
+     * @param filePath the file path to check.
+     */
+    private boolean checkFile(Path filePath) {
+        return Files.exists(filePath) && filePath.toFile().isFile();
+    }
 
     @Override
     protected void doInitialize(ZookeeperContext context) throws DubboTestException {
+        // checks the zookeeper binary file exists or not
+        if (checkFile(context.getSourceFile())) {
+            return;
+        }
         String zookeeperFileName = String.format(ZOOKEEPER_FILE_NAME_FORMAT, context.getVersion());
+        Path temporaryFilePath;
         try {
-            context.setSourceFile(Paths.get(Files.createTempDirectory("").getParent().toString(),
-                TEMPORARY_DIRECTORY_NAME,
-                zookeeperFileName));
+            temporaryFilePath = Paths.get(Files.createTempDirectory("").getParent().toString(),
+                TEMPORARY_DIRECTORY,
+                zookeeperFileName);
         } catch (IOException e) {
-            throw new RuntimeException(String.format("Cannot create the temporary directory, related directory:%s/%s",
-                TEMPORARY_DIRECTORY_NAME, zookeeperFileName), e);
-        }
-        // check if the zookeeper binary file exists
-        if (context.getSourceFile() != null && context.getSourceFile().toFile().isFile()) {
-            return;
+            throw new RuntimeException(String.format("Cannot create the temporary directory, file path: %s", TEMPORARY_DIRECTORY), e);
         }
+
         // create the temporary directory path.
+        try {
+            Files.createDirectories(temporaryFilePath.getParent());
+        } catch (IOException e) {
+            throw new RuntimeException(String.format("Failed to create the temporary directory to save zookeeper binary file, file path:%s", temporaryFilePath.getParent()), e);
+        }
+
+        // download zookeeper binary file in temporary directory.
+        String zookeeperBinaryUrl = String.format(ZOOKEEPER_BINARY_URL_FORMAT, context.getVersion(), context.getVersion());
+        try {
+            logger.info("It is beginning to download the zookeeper binary archive, it will take several minutes..." +
+                "\nThe zookeeper binary archive file will be download from " + zookeeperBinaryUrl + "," +
+                "\nwhich will be saved in " + temporaryFilePath.toString() + "," +
+                "\nalso it will be renamed to 'apache-zookeeper-bin.tar.gz' and moved into {project.dir}.tmp/zookeeper directory.\n");
+            URL url = new URL(zookeeperBinaryUrl);
+            HttpURLConnection connection = (HttpURLConnection) url.openConnection();
+            // set timeout when download the zookeeper binary archive file.
+            connection.setConnectTimeout(CONNECT_TIMEOUT);
+            // set timeout when read downloaded input stream to save in temporary file path.
+            connection.setReadTimeout(READ_TIMEOUT);
```

Comment:
> If the timeout expires before there is data available for read, a java.net.SocketTimeoutException is raised

The read timeout will not interrupt the downloading if the file data is received slowly and continuously. It is more reliable to check the download timeout in a separate checking thread, or use the request timeout of the asynchronous http client :  https://stackoverflow.com/a/38273306

Reply:
I have used `AsyncHttpClient` to download zookeeper binary archive file.